File: cpproxy/buffer.cpp

#include "buffer.h"

Buffer::Buffer() : m_current(0)
{
}

Buffer::Buffer(const uint8_t *data, unsigned int len) : m_current(0)
{
	write(data, len);
}

void Buffer::write(const uint8_t *data, unsigned int len)
{
	unsigned int oldSize = m_data.size();
	m_data.resize(oldSize + len);

	for(unsigned int i = 0; i < len; ++i)
		m_data[oldSize + i] = data[i];
}

void Buffer::read(uint8_t *data, unsigned int len)
{
	for(unsigned int i = 0; i < len; ++i)
		data[i] = m_data[m_current++];
}


void Buffer::peek(uint8_t *data, unsigned int len) const
{
	for(unsigned int i = 0; i < len; ++i)
		data[i] = m_data[m_current + i];
}


unsigned int Buffer::getSize() const
{
	return m_data.size() - m_current;
}

File: test_client/Buffer.java

public class Buffer
{
	private byte[] m_data;
	
	public Buffer()
	{
		m_data = new byte[0];
	}
	
	public void addData(byte[] array, int length)
	{
		byte[] data = new byte[m_data.length + length];
		
		for(int i = 0; i < m_data.length; ++i)
			data[i] = m_data[i];
		
		for(int i = 0; i < length; ++i)
			data[m_data.length + i] = array[i];
		
		m_data = data;
	}

	public void addData(byte b)
	{
		byte[] data = new byte[m_data.length + 1];
		
		for(int i = 0; i < m_data.length; ++i)
			data[i] = m_data[i];
		
		data[m_data.length] = b;
		
		m_data = data;
	}
	
	public byte[] getByteArray()
	{
		return m_data;
	}
}
